Migrate nuPickers to v8

bielu opened the pull request:
It is an initial version for v8, few things have to be still polished, but works! I am happy to continue to work on UI changes and/code optimizations. I think it is ready to be published as prerelease and check with other Umbraco developers if everything is working as intended

Commit:
Initial commit for v8

Commit:
Move events to components, migrate shared logic

Commit:
ignore rider files, migrate editor to v8

Commit:
Migrated enumpicker

Commit:
First changes for EnumCheckboxPicker

Commit:
Enum Labels Editor

Commit:
Enum Lables Editor, Enum PrefetchList Picker

Commit:
MIgrate DonetPagedListPicker, move unit testing to Nunit

Commit:
JsonLabels and Jsondropdown

Commit:
Lucene Picker

Commit:
LuceneLabels

Commit:
enum,json, lucene pickers

Commit:
bum minimal umbraco version to 8.1.1.

Commit:
correct building process, working on http handler

Commit:
restore original file extensions

Commit:
Solve problem with embed resources

Commit:
sql dropdown picker

Commit:
DotNet Pickers

Commit:
get list  of all searchers

Commit:
adjust namespace, refactor from propertyeditor to dataeditor

Commit:
Migrate next DataEditors, refactor names,

Commit:
migrate next DataType Editors, refactor names

Commit:
Json type a head data editor

Commit:
Lucene Checkbox, refactor name of lucene, sql picker

Commit:
Lucene Prefetch editor, replace PropertyEditorConstants with DataEditorConstants

Commit:
adjust namespace in not included editors yet

Commit:
Mass replace PropertyEditor with dataeditor, migrate LuceneRadioButton

Commit:
Replace PreValueField, prevalueeditor with configuration, migrate Relation Label

Commit:
Mass add name of Fields, unification of names

Commit:
Correct earlier Mass replacement

Commit:
Migrate sqlcheckboxpicker

Commit:
migrate sql labels

Commit:
migrate last sql pickers

Commit:
Use correct Configuration

Commit:
Restore original building script

Commit:
remove idea rider files

ronaldbarendse commented:
I see the `PickerPropertyValueConverter` still uses the `AssignedContentItem` to determine the `contextId` and `parentId`, but Umbraco 8 now provides the owner `IPublishedElement`:

https://github.com/uComponents/nuPickers/blob/218212a98982944ab85a6ff820a69958daafbcf0/source/nuPickers/PickerPropertyValueConverter.cs#L84-L85

bielu commented:
@ronaldbarendse  actually good catch, it was for now done to works, now I am doing additional refactor but I make PR with a version which is already working :)

Commit:
use owner IPublishedContent instead of assign content

bielu commented:
@ronaldbarendse moved to use owner :)

Commit:
Fix for empty prevalues

bowserm commented:
@bielu,

Any chance we could change the `GetPropertyValueType` method in the `PickerPropertyValueConverter` to the following? Right now, the property value converter in this pull request sets the type to `string`. This affects how ModelsBuilder generates the nupicker properties on its models.

    public override Type GetPropertyValueType(IPublishedPropertyType propertyType) => typeof(Picker);

I forked off of your `bielu/v8/develop` branch and tried it out locally and confirmed that ModelsBuilder generates the nuPicker properties with type `nuPickers.Picker` after this change.

Thanks for all the hard work on this!

bielu commented:
I will look at that's tomorrow morning 😜

Commit:
Change Value Convert to use Picker instead of String

bielu commented:
@bowserm changed :)

ismailmayat committed:
remove notimplemented which is currently generating errors in log

joepvtl commented:
Hey @bielu , the saveFormat of the [EnumPrefetchListPicker](https://github.com/bielu/nuPickers/blob/v8/develop/source/nuPickers/DataEditors/EnumPrefetchListPicker/EnumPrefetchListPickerConfigurationEditor.cs) isn't getting saved. This is because the value is not being assigned.

-Joep

bielu commented:
@joepvtl I will look into that tomorrow morning :)

bowserm commented:
I am seeing what I believe is the saveFormat error on the [EnumRadioButtonPicker](https://github.com/bielu/nuPickers/blob/v8/develop/source/nuPickers/DataEditors/EnumRadioButtonPicker/EnumRadioButtonPickerConfigurationEditor.cs). When I look in the DB, this is what I am seeing as the prevalues for my datatype when I selected csv as the saveFormat:

```
{"useLabel":false,"dataSource":{"assemblyName":"ScientificGames.Core.dll","apiController":"EnumDataSourceApi","enumName":"ScientificGames.Core.Models.DataTypeDataSources.LeftRightTopBottomOrientation"},"customLabel":null,"layoutDirection":"horizontal","saveFormat":null,"items":[]}
```

I'll race you to a fix blielu :). I'll see if the same problem is happening to other editors.

bowserm committed:
Ensure that saveFormat is saved as a string for every property editor

bowserm commented:
@bielu I made a PR to the `v8/develop` branch on your fork. I think I have it working.

bielu committed:
Merge pull request #1 from bowserm/v8/develop

Ensure that saveFormat is saved as a string for every property editor

bielu commented:
@bowserm thanks for the help, I reviewed that and merged code.
Actually I am quite little less active now as I have crazy time but I am happy to merge PRs.

PezCo commented:
Is this Likely to be the definitive upgrade path for nuPickers? im a bit confused between this and https://github.com/uComponents/Pikachu which says its a v8 rebuild.

is there any "official" word on plans for v8 its almost a year old now.

bielu commented:
@PezCo this is Upgrade of actual nuPickers which I started, as I needed use nuPickers in one of the projects, Pikachu was supposed to rebuild that for v8, not direct migration like that branch. @Hendy is responsible for the official part of that, but not sure what is his decision at all :).

ronaldbarendse commented:
If you need an alternative to nuPickers for Umbraco 8, you might also be interested in Data List from @leekelleher: https://github.com/leekelleher/umbraco-contentment/blob/master/docs/editors/data-list.md.

PezCo commented:
Ideally i want this branch.
I have a fairly large public Umbraco v7 site that i'm upgrading to v8 that makes use of nuPickers.
Unfortunately im not really experienced in umbraco plugins or its source code so not in much of a position to help here, im just looking for the least pain free way to migrate.

Commit:
upgraded to latest version of Umbraco + fixed issue with null check on render macro by injecting UmbracoComponentRenderer

Commit:
Merge pull request #2 from Ric-Carey/v8/develop

upgraded to latest version of Umbraco + fixed issue with null check o…

jesperbrasmussen commented:
Great work. I have made my own build of your branch to use on an Umbraco V8 installation (v. 8.6.1). But i ran into some problems with PanicExceptions getting thrown in PickerPropertyValueConverter when it tries to access the parent property. 
It looks like it's related to NuCache, but i can't seem to solve it by just rebuilding the cache.
I'm using modelsbuilder and it happens when i try to get the values of the property that use the nupicker datatype. It's shows op fine in the Backoffice.

```
public override object ConvertIntermediateToObject(IPublishedElement owner, IPublishedPropertyType propertyType,
            PropertyCacheLevel referenceCacheLevel, object inter, bool preview)
        {
            int contextId = -1;
            int parentId = -1;

            IPublishedContent assignedContentItem = owner as IPublishedContent;

            if (assignedContentItem != null)
            {
                contextId = assignedContentItem.Id;

                if (assignedContentItem.Parent != null)
                {
                    parentId = assignedContentItem.Parent.Id;
                }
            }

            return new Picker(
                contextId,
                parentId,
                propertyType.EditorAlias,
                propertyType.DataType.Id,
                propertyType.Alias,
                inter);
        }
```


My workaround was to surround the failing code with a try/cacth and ignore the PanicException. It solves it in my case, but i have yet to discover if it has any sideeffects.

```
if (assignedContentItem != null)
            {
                contextId = assignedContentItem.Id;

                try
                {
                    if (assignedContentItem.Parent != null)
                    {
                        parentId = assignedContentItem.Parent.Id;
                    }
                }
                catch (PanicException e)
                {
                    //Ignore PanicException
                }
                
            }
```

bielu commented:
@jesperbrasmussen can you add which version of v8 are you using? and I will look into that issue when I will have time

jesperbrasmussen commented:
> @jesperbrasmussen can you add which version of v8 are you using? and I will look into that issue when I will have time

Yes it's v. 8.6.1

bielu commented:
@jesperbrasmussen can you also paste  exception log as maybe I will be find out that just by exception as I check Umbraco source code and not much change here about  ConvertIntermediateToObject and I will need to recreate but would prefer be sure it is same exception if I will cause any :)

jesperbrasmussen commented:
I have just tested it on the lastest version of Umbraco which is v8.6.2, and I still get get the same error. 

This is the first part of the exception:

```
<Error>
<Message>An error has occurred.</Message>
<ExceptionMessage>invalid item type</ExceptionMessage>
<ExceptionType>Umbraco.Core.Exceptions.PanicException</ExceptionType>
<StackTrace> at Umbraco.Web.PublishedCache.NuCache.PublishedContent.GetGetterById() in D:\a\1\s\src\Umbraco.Web\PublishedCache\NuCache\PublishedContent.cs:line 123 at Umbraco.Web.PublishedCache.NuCache.PublishedContent.get_Parent() in D:\a\1\s\src\Umbraco.Web\PublishedCache\NuCache\PublishedContent.cs:line 261 at nuPickers.PickerPropertyValueConverter.ConvertIntermediateToObject(IPublishedElement owner, IPublishedPropertyType propertyType, PropertyCacheLevel referenceCacheLevel, Object inter, Boolean preview) in S:\Solutions (external)\nuPickers\source\nuPickers\PickerPropertyValueConverter.cs:line 118 at Umbraco.Core.Models.PublishedContent.PublishedPropertyType.ConvertInterToObject(IPublishedElement owner, PropertyCacheLevel referenceCacheLevel, Object inter, Boolean preview) in D:\a\1\s\src\Umbraco.Core\Models\PublishedContent\PublishedPropertyType.cs:line 219 at Umbraco.Web.PublishedCache.NuCache.Property.GetValue(String culture, String segment) in D:\a\1\s\src\Umbraco.Web\PublishedCache\NuCache\Property.cs:line 211 at Umbraco.Web.PublishedPropertyExtension.Value[T](IPublishedProperty property, String culture, String segment, Fallback fallback, T defaultValue) in D:\a\1\s\src\Umbraco.Web\PublishedPropertyExtension.cs:line 39 at Umbraco.Web.PublishedContentExtensions.Value[T](IPublishedContent content, String alias, String culture, String segment, Fallback fallback, T defaultValue) in D:\a\1\s\src\Umbraco.Web\PublishedContentExtensions.cs:line 163 at MyProject.UmbracO2.Models.CustomerMember.get_AllowedTypesForMember() in S:\Solutions\MyProject\MyProject.com_v8\MyProject.UmbracO2\Models\CustomerMember.generated.cs:line 60 at MyProject.UmbracO2.Models.CustomerMember.MyProject.Core.Interfaces.Member.ICustomerMember.get_AllowedTypesForMember() in S:\Solutions\MyProject\MyProject.com_v8\MyProject.UmbracO2\Models\CustomerMember.cs:line 16 at MyProject.UmbracO2.Services.UmbracoMembershipService.GetAllowedTypes(Int32 memberId) in S:\Solutions\MyProject\MyProject.com_v8\MyProject.UmbracO2\Services\UmbracoMembershipService.cs:line 63 at MyProject.UmbracO2.Services.Member.MemberAdministrationService.GetAllowedTypes(Int32 memberId) in S:\Solutions\MyProject\MyProject.com_v8\MyProject.UmbracO2\Services\Member\MemberAdministrationService.cs:line 347 at MyProject.UmbracO2.Controllers.Backoffice.Section.AdminToolController.<>c__DisplayClass6_0.<ListUsers>b__3(IMember group) in S:\Solutions\MyProject\MyProject.com_v8\MyProject.UmbracO2\Controllers\Backoffice\Section\AdminToolController.cs:line 71 at System.Linq.Lookup`2.Create[TSource](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer) at System.Linq.Enumerable.ToLookup[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector) at MyProject.UmbracO2.Controllers.Backoffice.Section.AdminToolController.ListUsers(Boolean includeInactive) in 
```

bielu commented:
@jesperbrasmussen thanks that error log is helpful! Will look over weekend how to fix that issue! :)

bielu commented:
@jesperbrasmussen I tried on clear umbraco 8.6.2 today and can't reproduce that issue,

jesperbrasmussen commented:
> @jesperbrasmussen I tried on clear umbraco 8.6.2 today and can't reproduce that issue,

Ok. I'll try to find time to do a test on a clean umbraco 8.6.2, and see if i can find a way to reproduce the error

niekvanderreest committed:
Fixed issue for nuPickers added to Membertypes

When a nuPicker is added to a custum MemberType and an instcance of that membertype is loaded through use of the memberservice i.e. Umbraco.Web.Composing.Current.UmbracoHelper.Member(id), the PickerPropertyValueConverter throws an error on line: 117.

The check assignedContentItem.Parent != null appears not to be a valid check for IMember eventhough it's successfully cast to IPublishedContent, the fix works around the check by getting the parent id from the Path property.

Commit:
Merge pull request #3 from niekvanderreest/patch-1

Fixed issue for nuPickers added to Membertypes
